plugins/ltac/g_rewrite.mlg

```diff
@@ -99,30 +99,39 @@ ARGUMENT EXTEND rewstrategy
 
     RAW_PRINTED BY { pr_raw_strategy env sigma }
     GLOB_PRINTED BY { pr_glob_strategy env sigma }
+END
 
-  | [ glob(c) ] -> { StratConstr (c, true) }
-  | [ "<-" constr(c) ] -> { StratConstr (c, false) }
-  | [ "subterms" rewstrategy(h) ] -> { StratUnary (Subterms, h) }
-  | [ "subterm" rewstrategy(h) ] -> { StratUnary (Subterm, h) }
-  | [ "innermost" rewstrategy(h) ] -> { StratUnary(Innermost, h) }
-  | [ "outermost" rewstrategy(h) ] -> { StratUnary(Outermost, h) }
-  | [ "bottomup" rewstrategy(h) ] -> { StratUnary(Bottomup, h) }
-  | [ "topdown" rewstrategy(h) ] -> { StratUnary(Topdown, h) }
-  | [ "id" ] -> { StratId }
-  | [ "fail" ] -> { StratFail }
-  | [ "refl" ] -> { StratRefl }
-  | [ "progress" rewstrategy(h) ] -> { StratUnary (Progress, h) }
-  | [ "try" rewstrategy(h) ] -> { StratUnary (Try, h) }
-  | [ "any" rewstrategy(h) ] -> { StratUnary (Any, h) }
-  | [ "repeat" rewstrategy(h) ] -> { StratUnary (Repeat, h) }
-  | [ rewstrategy(h) ";" rewstrategy(h') ] -> { StratBinary (Compose, h, h') }
-  | [ "(" rewstrategy(h) ")" ] -> { h }
-  | [ "choice" ne_rewstrategy_list(h) ] -> { StratNAry (Choice, h) }
-  | [ "old_hints" preident(h) ] -> { StratHints (true, h) }
-  | [ "hints" preident(h) ] -> { StratHints (false, h) }
-  | [ "terms" constr_list(h) ] -> { StratTerms h }
-  | [ "eval" red_expr(r) ] -> { StratEval r }
-  | [ "fold" constr(c) ] -> { StratFold c }
+GRAMMAR EXTEND Gram
+  GLOBAL: rewstrategy;
+  rewstrategy:
+    [ LEFTA
+      [ h = SELF; ";"; h' = rewstrategy_atom -> { StratBinary (Compose, h, h') } ] ]
+  ;
+  rewstrategy_atom:
+    [ RIGHTA
+      [ c = glob -> { StratConstr (c, true) }
+      | "<-"; c = constr -> { StratConstr (c, false) }
+      | IDENT "subterms"; h = SELF -> { StratUnary (Subterms, h) }
+      | IDENT "subterm"; h = SELF -> { StratUnary (Subterm, h) }
+      | IDENT "innermost"; h = SELF -> { StratUnary(Innermost, h) }
+      | IDENT "outermost"; h = SELF -> { StratUnary(Outermost, h) }
+      | IDENT "bottomup"; h = SELF -> { StratUnary(Bottomup, h) }
+      | IDENT "topdown"; h = SELF -> { StratUnary(Topdown, h) }
+      | IDENT "id" -> { StratId }
+      | IDENT "fail" -> { StratFail }
+      | IDENT "refl" -> { StratRefl }
+      | IDENT "progress"; h = SELF -> { StratUnary (Progress, h) }
+      | IDENT "try"; h = SELF -> { StratUnary (Try, h) }
+      | IDENT "any"; h = SELF -> { StratUnary (Any, h) }
+      | IDENT "repeat"; h = SELF -> { StratUnary (Repeat, h) }
+      | IDENT "choice"; h = LIST1 rewstrategy -> { StratNAry (Choice, h) }
```

Comment:
Also `choice` is a bit odd, for example `choice fail fail ; fail`, which means `choice fail (fail ; fail)`.  And `rewstrategy_atom` seems maybe odd since `choice` contains the more-general `rewstrategy`.

Or could the production be `IDENT "choice" LIST1 SELF`?

Reply:
The syntax and precedence of `choice` is not very intuitive. E.g., an alternative could have been `[ rewstrategy | ... | rewstrategy ]`.
We may try to change it (at least to `IDENT "choice"; h = LIST1 rewstrategy_atom`), if you have the courage to test it on the CI.

Reply:
> Elsewhere in the doc we've often used the plural form to refer to lists of items

If you think it is more consistent with the rest of the refman, why not. But that would also mean changing the code which uses everywhere `rewstrategy` for the sequence.

Reply:
_Au contraire, mon ami,_ no need to change the code.  Just add `RENAME rewstrategy rewstrategies` in `common.edit_mlg` to change it in the doc only.

> We may try to change it (at least to IDENT "choice"; h = LIST1 rewstrategy_atom), if you have the courage to test it on the CI.

Courage??  I don't want to try that myself, but it seems an easy experiment.

Reply:
PS  You may need to change a few references in the documentation text from `rewstrategy` to `rewstrategies` so it matches the syntax.

Reply:
I think I'm a bit afraid of the changes it would imply. I'd prefer you take the responsibility of a RENAME.

Reply:
OK, but please get the grammar right and the refman to build first.

Reply:
I think choice should be just `| IDENT "choice"; SELF`.  If the user wants to specify multiple rewstrategy_atoms they must parenthesize them and separate with semicolons.  If there's only one, no parens needed.  This avoids adding another semicolon ambiguity.